### PR TITLE
feat(dashboardeditor): support editable title widgets natively

### PR DESCRIPTION
**Pull request by scottdickerson**
Closes #

**Summary**

- feat(dashboardeditor): currently all the edit state behavior has to be kept in the parent widget.  This moves the state for the edit title capability into the dashboardeditor
- Improve the PageTitleBar to allow custom titles to be rendererd

**Change List (commits, features, bugs, etc)**

- feat(DashboardEditorHeader): add state tracking and widgets to edit the dashboard title
- feat(PageTitleBar): allow passing of a customer renderTitleFunction so alternate nodes can be returned than the default title string

**Acceptance Test (how to verify the PR)**

- Verify the custom render function in the pagetitlebar stories
- Test DashboardEditor story for editing the dashboard title

**Regression Test (how to make sure this PR doesn't break old functionality)**

- Verify that normal pagetitlebar title is non-editable, verify if a dashboardeditor doesn't have a onEditTitle handler that it doesn't render the edit button

### CODE REVIEW

**Review by kevinsperrine — APPROVED** [review]

**Comment by bjornalm** [comment]
which github issue is this for?

**Comment by bjornalm** [comment]
image
image


I think we are missing spacing between the input field and the cancel button. There should probably also be spacing between the title and the edit button.

File: packages/react/src/components/DashboardEditor/DashboardEditorHeader/DashboardEditorHeader.jsx [hunk 1]
```diff
@@ -220,13 +234,61 @@ const DashboardEditorHeader = ({
     </div>
   );
 
+  // handle the edit title button
+  const handleEditClick = useCallback(() => {
+    setIsTitleEditMode(true);
+  }, []);
+
+  const editWidgets = useMemo(
```
Comment (bjornalm) [inline packages/react/src/components/DashboardEditor/DashboardEditorHeader/DashboardEditorHeader.jsx:242]: nit: I don't think the input and buttons are considered widgets

File: packages/react/src/components/PageTitleBar/PageTitleBar.story.jsx [hunk 2]
```diff
@@ -438,6 +442,23 @@ WithCondensedHeader.story = {
   name: 'with condensed with primary + secondary buttons',
 };
 
+export const CustomRenderFunction = () => (
```
Comment (bjornalm) [inline packages/react/src/components/PageTitleBar/PageTitleBar.story.jsx:445]: call it withCustomRenderFunction to be consistent?

**Review by bjornalm — APPROVED** [review]
Nice work

File: packages/react/src/components/PageTitleBar/PageTitleBar.story.jsx (hunk 2 above)
Comment (davidicus) [inline packages/react/src/components/PageTitleBar/PageTitleBar.story.jsx:445]:
```suggestion
export const WithCustomRenderFunction = () => (
```

**Review by davidicus — APPROVED** [review]

**Review by herleraja — APPROVED** [review]
LGTM, but tests are failing.

File: packages/react/src/components/DashboardEditor/DashboardEditorHeader/DashboardEditorHeader.jsx (hunk 1 above)
Comment (scottdickerson) [reply packages/react/src/components/DashboardEditor/DashboardEditorHeader/DashboardEditorHeader.jsx:242]: thanks, will rename

Comment (scottdickerson) [reply packages/react/src/components/DashboardEditor/DashboardEditorHeader/DashboardEditorHeader.jsx]: thanks!

**Pull request merged**